### ISSUE TITLE
Compilation error about -fPIC

### ISSUE THREAD

**Issue by chunjiangzhu** [CannyLab/tsne-cuda#19]
First thanks a lot for providing the code, and suggestions on our previous fail on parallel_for #11 !
Compiling from source Error:

> [ 38%] Building CXX object CMakeFiles/tsnecuda.dir/third_party/faiss/gpu/GpuAutoTune.cpp.o
> Linking CXX shared library libtsnecuda.so
> /usr/bin/ld: CMakeFiles/tsnecuda.dir/./tsnecuda_intermediate_link.o: relocation R_X86_64_32S against `__nv_module_id' can not be used when making a shared object; recompile with -fPIC
> CMakeFiles/tsnecuda.dir/./tsnecuda_intermediate_link.o: error adding symbols: Bad value
> collect2: error: ld returned 1 exit status
> make[2]: *** [libtsnecuda.so] Error 1
> make[1]: *** [CMakeFiles/tsnecuda.dir/all] Error 2
> make: *** [all] Error 2

Gcc version:

> gcc (Ubuntu 4.8.5-4ubuntu8~14.04.2) 4.8.5
> Copyright (C) 2015 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

GPU:

> ./deviceQuery Starting...
> 
> CUDA Device Query (Runtime API) version (CUDART static linking)
> 
> Detected 2 CUDA Capable device(s)
> 
> Device 0: "Tesla K40c"
> CUDA Driver Version / Runtime Version 9.0 / 9.0
> CUDA Capability Major/Minor version number: 3.5
> Total amount of global memory: 11440 MBytes (11995578368 bytes)
> (15) Multiprocessors, (192) CUDA Cores/MP: 2880 CUDA Cores
> GPU Max Clock rate: 745 MHz (0.75 GHz)
> Memory Clock rate: 3004 Mhz
> Memory Bus Width: 384-bit
> L2 Cache Size: 1572864 bytes
> Maximum Texture Dimension Size (x,y,z) 1D=(65536), 2D=(65536, 65536), 3D=(4096, 4096, 4096)
> Maximum Layered 1D Texture Size, (num) layers 1D=(16384), 2048 layers
> Maximum Layered 2D Texture Size, (num) layers 2D=(16384, 16384), 2048 layers
> Total amount of constant memory: 65536 bytes
> Total amount of shared memory per block: 49152 bytes
> Total number of registers available per block: 65536
> Warp size: 32
> Maximum number of threads per multiprocessor: 2048
> Maximum number of threads per block: 1024
> Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
> Max dimension size of a grid size (x,y,z): (2147483647, 65535, 65535)
> Maximum memory pitch: 2147483647 bytes
> Texture alignment: 512 bytes
> Concurrent copy and kernel execution: Yes with 2 copy engine(s)
> Run time limit on kernels: No
> Integrated GPU sharing Host Memory: No
> Support host page-locked memory mapping: Yes
> Alignment requirement for Surfaces: Yes
> Device has ECC support: Enabled
> Device supports Unified Addressing (UVA): Yes
> Supports Cooperative Kernel Launch: No
> Supports MultiDevice Co-op Kernel Launch: No
> Device PCI Domain ID / Bus ID / location ID: 0 / 2 / 0
> Compute Mode:
> < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >
> 
> Device 1: "Tesla K40c"
> CUDA Driver Version / Runtime Version 9.0 / 9.0
> CUDA Capability Major/Minor version number: 3.5
> Total amount of global memory: 11440 MBytes (11995578368 bytes)
> (15) Multiprocessors, (192) CUDA Cores/MP: 2880 CUDA Cores
> GPU Max Clock rate: 745 MHz (0.75 GHz)
> Memory Clock rate: 3004 Mhz
> Memory Bus Width: 384-bit
> L2 Cache Size: 1572864 bytes
> Maximum Texture Dimension Size (x,y,z) 1D=(65536), 2D=(65536, 65536), 3D=(4096, 4096, 4096)
> Maximum Layered 1D Texture Size, (num) layers 1D=(16384), 2048 layers
> Maximum Layered 2D Texture Size, (num) layers 2D=(16384, 16384), 2048 layers
> Total amount of constant memory: 65536 bytes
> Total amount of shared memory per block: 49152 bytes
> Total number of registers available per block: 65536
> Warp size: 32
> Maximum number of threads per multiprocessor: 2048
> Maximum number of threads per block: 1024
> Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
> Max dimension size of a grid size (x,y,z): (2147483647, 65535, 65535)
> Maximum memory pitch: 2147483647 bytes
> Texture alignment: 512 bytes
> Concurrent copy and kernel execution: Yes with 2 copy engine(s)
> Run time limit on kernels: No
> Integrated GPU sharing Host Memory: No
> Support host page-locked memory mapping: Yes
> Alignment requirement for Surfaces: Yes
> Device has ECC support: Enabled
> Device supports Unified Addressing (UVA): Yes
> Supports Cooperative Kernel Launch: No
> Supports MultiDevice Co-op Kernel Launch: No
> Device PCI Domain ID / Bus ID / location ID: 0 / 4 / 0
> Compute Mode:
> < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >
> 
> Peer access from Tesla K40c (GPU0) -> Tesla K40c (GPU1) : Yes
> Peer access from Tesla K40c (GPU1) -> Tesla K40c (GPU0) : Yes
> deviceQuery, CUDA Driver = CUDART, CUDA Driver Version = 9.0, CUDA Runtime Version = 9.0, NumDevs = 2
> Result = PASS

CUDA:

> nvcc: NVIDIA (R) Cuda compiler driver
> Copyright (c) 2005-2017 NVIDIA Corporation
> Built on Fri_Sep__1_21:08:03_CDT_2017
> Cuda compilation tools, release 9.0, V9.0.176

**Comment by DavidMChan** [comment]
Hmm. This is interesting. Can you add the output of `make VERBOSE=1` to a GIST and share it? This will give the exact compiler options which are invoked when compiling, and could help us track down the issue. It seems like the -fPIC flag is missing - but it should be passed in on line 65 of the CMakeLists.txt file.

**Comment by chunjiangzhu** [comment]
> Hmm. This is interesting. Can you add the output of `make VERBOSE=1` to a GIST and share it? This will give the exact compiler options which are invoked when compiling, and could help us track down the issue. It seems like the -fPIC flag is missing - but it should be passed in on line 65 of the CMakeLists.txt file.

Yes, please take a look. Not sure whether I did what you expected. But let me know if I need to provide more info. Thanks!
https://gist.github.com/chunjiangzhu/fc35f4d93937979d954a17fb571bc369

**Comment by DavidMChan** [comment]
Thanks! That's almost what I needed - however because you had already built the code, the output is unfortunately a bit uninformative. Can you re-run with `make -B VERBOSE=1` which should re-build the project from scratch. There's going to be a lot of output text, so perhaps running `make -B VERBOSE=1 &> output.txt` might be a good option.

**Comment by chunjiangzhu** [comment]
Hi David, thanks a lot for the guidance! Please see the new gist.
https://gist.github.com/chunjiangzhu/c725e5bf4fe5129a8742971b6c1508e7

**Comment by chunjiangzhu** [comment]
Hi, we are still unsuccessful in the compilation:(. We are thinking if it is convenient for you compile the code with "-gencode=arch=compute_30,code=sm_30" or "-gencode=arch=compute_35,code=sm_35" on line 59 or 60 of the CMakeLists.txt file, and then somehow transfer the compiled files to us. Thanks!

**Comment by rmrao** [comment]
We just added conda support. You should be able to download with `conda install tsnecuda -c cannylab`, assuming you have CUDA 9.0 installed. Hopefully this fixes your compilation issues.
